RA-23

KinCadeFoster opened the pull request:
- Из фрагмента со списком рецептов передавал только recipeId (вместо объекта рецепта целиком).
- Внутри ViewModel создал функцию loadRecipe().
- Добавил комментарий TODO с сообщением 'load from network'. -- Метод вызывается один раз при создании фрагмента и служит для первичной инициализации стейта. -- Метод принимает id рецепта и присваивает полученный из хранилища рецепт соответствующему свойству стейта RecipeState.
- Перенес метод getFavorites() во ViewModel для первичной инициализации поля стейта isFavorite актуальным значением.
- В методе loadRecipe() при создании объекта стейта: Проинициализировал свойство стейта isFavorite, используя getFavorites().
- Проинициализировал свойство стейта portionsCount, используя значение текущего стейта.
- Перенес подписку на изменения LiveData в метод initUI(). Завернул все содержимое метода в лямбду наблюдателя и доставать все необходимые значения из стейта.
- Вынес логику проверки рецепта на избранность: -- Внутри ViewModel создал метод onFavoritesClicked() и вызываю его при клике на иконку Favorites. Вынес логику определения из прослушивателя кликов и в зависимости от результата проверки обновляю поле стейта isFavorite. При присвоении нового стейта переменной LiveData не создаваю новый объект, а копирую предыдущий стейт с изменением нужного свойства.

KinCadeFoster committed:
RA-23
- Из фрагмента со списком рецептов передавал только recipeId (вместо объекта рецепта целиком).
- Внутри ViewModel создал функцию loadRecipe().
- Добавил комментарий TODO с сообщением 'load from network'.
-- Метод вызывается один раз при создании фрагмента и служит для первичной инициализации стейта.
-- Метод принимает id рецепта и присваивает полученный из хранилища рецепт соответствующему свойству стейта RecipeState.
- Перенес метод getFavorites() во ViewModel для первичной инициализации поля стейта isFavorite актуальным значением.
- В методе loadRecipe() при создании объекта стейта:
Проинициализировал свойство стейта isFavorite, используя getFavorites().
- Проинициализировал свойство стейта portionsCount, используя значение текущего стейта.
- Перенес подписку на изменения LiveData в метод initUI(). Завернул все содержимое метода в лямбду наблюдателя и доставать все необходимые значения из стейта.
- Вынес логику проверки рецепта на избранность:
-- Внутри ViewModel создал метод onFavoritesClicked() и вызываю его при клике на иконку Favorites. Вынес логику определения из прослушивателя кликов и в зависимости от результата проверки обновляю поле стейта isFavorite. При присвоении нового стейта переменной LiveData не создаваю новый объект, а копирую предыдущий стейт с изменением нужного свойства.

aGroupoid commented on app/src/main/java/ru/example/recipeapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -29,13 +26,12 @@ class RecipeFragment : Fragment() {
+        arguments?.getInt(Constants.ARG_RECIPE)
+    }
+    private val viewModel: RecipeViewModel by viewModels {
+        RecipeViewModelFactory(requireContext(), recipeId!!)
```
Использовать !! не рекомендуется, лучше передавать nullable recipeId, на стороне viewModel этот  null обрабатывать. Сейчас при неправильном стейте  просто приложение упадёт

aGroupoid commented on app/src/main/java/ru/example/recipeapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -29,13 +26,12 @@ class RecipeFragment : Fragment() {
-    private val sharedPrefs by lazy {
-        requireContext().getSharedPreferences(SHARED_PREFS_NAME, android.content.Context.MODE_PRIVATE)
+    private val recipeId by lazy {
+        arguments?.getInt(Constants.ARG_RECIPE)
```
На каждом экране сейчас показывается классический бургер, ты кладешь Parcelable Recipe, а пытаешься поулчить int, в итоге возващается 0 - id классического бургера.
Дублируется логика из getRecipeFromArguments

aGroupoid commented on app/src/main/java/ru/example/recipeapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -47,12 +43,7 @@ class RecipeFragment : Fragment() {
-
         if (recipe != null) {
             initUI(recipe)
             initRecycler(recipe)
```
Все работы с recipe должны быть вынесены в observer под стейт.  initUI  и initRecycler должны работать без передачи recipe в функцию, всё что зависит от этого нужно перенести внутри observer

aGroupoid commented on app/src/main/java/ru/example/recipeapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -76,40 +67,26 @@ class RecipeFragment : Fragment() {
-
             binding.headerImage.setImageBitmap(imageResources)
             binding.headerImage.contentDescription = getString(
                 R.string.content_description_recipe_image, recipe.title
```
ссылаешься не на даныне из state

aGroupoid reviewed (changes requested):
При повороте экрана сбрасывается количество порций. В листенере seekBar неправильно менять текст, у нас единственный актуальный источник данных - стейт во вью моделе

KinCadeFoster committed:
RA-23
- Правки по замечаниям

aGroupoid commented on app/src/main/java/ru/example/recipeapp/ui/recipes/recipe/RecipeFragment.kt:
```diff
@@ -43,31 +49,21 @@ class RecipeFragment : Fragment() {
+        viewModel.recipeLiveData.observe(viewLifecycleOwner) { stateRecipe ->
+            stateRecipe.recipe?.let { recipe ->
+                initUI(stateRecipe)
+                initRecycler(recipe)
```
С такой реализаций появился забавный баг. Понажимай на иконку сердечка раз 10, посмотри как меняются твой адаптер. 
Я в прошлом комменте указал, что нужно перенести под observer только те строки, где участвует аргумент рецепта. Вся остальная функция initRecycler и initUi должна не зависеть от аргумента рецепта и вызываться единожды вне observer

aGroupoid replied:
Надо понять, что должно вызываться один раз, а что должно реагировать на изменения. То есть setOnClickListener достаточно устанавливать единожды, а не каждый раз при изменении стейта

KinCadeFoster committed:
RA-23
- Правки по замечаниям (баг с очередностью выполнения)
- Пофиксил баг, что после смены ориетации экрана, количество порции сбрасивалось на 1, добавил функцию updatePortions во вью модель

aGroupoid reviewed (approved):
Молодец! Всё верно
